unity/Assets/Scripts/ActionDispatcher.cs

```diff
@@ -0,0 +1,233 @@
+
+using System.Reflection;
+using System.Collections.Generic;
+using System;
+using UnityEngine;
+
+
+/*
+    The ActionDispatcher takes a dynamic object with an 'action' property and 
+    maps this to a method.  Matching is performed using the parameter names. 
+    In the case of method overloading, the best match is returned based on the
+    number of matched named parameters.  For a method to qualify for dispatching
+    it must be public and have a return type of void.  The following method 
+    definitions are permitted:
+
+    public void MoveAhead()
+    public void MoveAhead(string direction)
+    public void MoveAhead(ServerAction action)
+    public void MoveAhead(float moveMagnitude, rotation=0.0f)
+
+
+    Creating the following overloaded set of functions will not work as expected:
+
+    public void Teleport(int x, int y)
+    public void Teleport(int x, short y)
+
+    as well the following scenario should also be avoided:
+
+    public void ObjectVisible(bool foo, int x, int y)
+    public void ObjectVisible(int x, int y, bool foo)
+
+
+    The reason for the aforementioned restrictions is twofold, we pass the arguments to
+    Unity serialized using json.  This restricts the types that can be passed to
+    C# as well even if we serialized using a different format, Python does not 
+    have all the same primitives, such as 'short'.  Second, we allow actions
+    to be invoked from the Python side using keyword args which doesn't preserve order.
+
+    These restrictions shouldn't present themselves as creating duplicate public
+    actions with different orders, but identically named parameters would lead to
+    confusion should be avoided.
+*/
+public static class ActionDispatcher {
+    private static Dictionary<Type, Dictionary<string, List<MethodInfo>>> methodDispatchTable = new Dictionary<Type, Dictionary<string, List<MethodInfo>>>();
+
+
+    private static MethodInfo getDispatchMethod(System.Object target, dynamic serverCommand) {
+        Type t = target.GetType();
+        if (!methodDispatchTable.ContainsKey(t)) {
+            System.Reflection.MethodInfo[] allMethods = target.GetType().GetMethods(BindingFlags.Public | BindingFlags.Instance);
+            List<Type> hierarchy = new List<Type>();
+            // not completely generic
+            Type ht = t;
+            while (ht != typeof(object)) {
+                hierarchy.Add(ht);
+                ht = ht.BaseType;
+            }
+
+            Dictionary<string, List<MethodInfo>> methodDispatch = new Dictionary<string, List<MethodInfo>>();
```

Comment:
We might be able to cache this after an `Init` call when the runtime type of the agent changes. Just an optimization thought for the future, but we should do this first.